diff --git a/backend/genesys/tables_model.cpp b/backend/genesys/tables_model.cpp
--- a/backend/genesys/tables_model.cpp
+++ b/backend/genesys/tables_model.cpp
@@ -668,7 +668,7 @@ void genesys_init_usb_device_tables()
     model.x_offset = 2.2;
     model.y_offset = 9.0;
     model.x_size = 216.70;
-    model.y_size = 300.0;
+    model.y_size = 5000.0;
 
     model.y_offset_calib_white = 0.0;
     model.y_size_calib_mm = 3.0;
@@ -722,7 +722,7 @@ void genesys_init_usb_device_tables()
     model.x_offset = 0.0;
     model.y_offset = 8.0;
     model.x_size = 216.0;
-    model.y_size = 300.0;
+    model.y_size = 5000.0;
 
     model.y_offset_calib_white = 1.0;
     model.y_size_calib_mm = 3.0;
@@ -775,7 +775,7 @@ void genesys_init_usb_device_tables()
     model.x_offset = 2.1;
     model.y_offset = 8.7;
     model.x_size = 216.70;
-    model.y_size = 297.5;
+    model.y_size = 5000.0;
 
     model.y_offset_calib_white = 0.0;
     model.y_size_calib_mm = 3.0;
@@ -831,7 +831,7 @@ void genesys_init_usb_device_tables()
     model.x_offset = 2.1;
     model.y_offset = 8.7;
     model.x_size = 216.70;
-    model.y_size = 297.5;
+    model.y_size = 5000.0;
 
     model.y_offset_calib_white = 0.0;
     model.y_size_calib_mm = 3.0;
diff --git a/backend/genesys/genesys.cpp b/backend/genesys/genesys.cpp
--- a/backend/genesys/genesys.cpp
+++ b/backend/genesys/genesys.cpp
@@ -989,7 +989,7 @@ void scanner_move_back_home(Genesys_Device& dev, bool wait_until_home)
     }
 
     if (wait_until_home) {
-        for (unsigned i = 0; i < 300; ++i) {
+        for (unsigned i = 0; i < 1800; ++i) {
             auto status = scanner_read_status(dev);
 
             if (status.is_at_home) {
